src/sync.cpp

```diff
@@ -36,6 +36,401 @@ const int Sync::FILE_UPDATE_DELAY_DS = 30;
+        if (!ffp.isvalid)
+        {
+            LOG_err << "Invalid fingerprint";
+            return;
+        }
+    }
+    else if (fa.type == FOLDERNODE)
+    {
+        combinedFingerprint(ffp, fsaccess, paths);
+    }
+    else
+    {
+        assert(false && "Invalid node type");
+    }
+}
+
+struct FileFingerprintComparator
+{
+    bool operator()(const FileFingerprint& lhs, const FileFingerprint& rhs) const
+    {
+        return FileFingerprintCmp{}(&lhs, &rhs);
+    }
+};
+
+using FingerprintMap = std::multimap<FileFingerprint, LocalNode*, FileFingerprintComparator>;
+
+// Collects all LocalNodes by storing them in `fingerprints`, keyed by FileFingerprint
+void collectAllFingerprints(FingerprintMap& fingerprints, LocalNode& l)
+{
+    FileFingerprint ffp;
+    computeFingerprint(ffp, l);
+    if (ffp.isvalid)
+    {
+        fingerprints.insert(std::make_pair(ffp, &l));
+    }
+    if (l.type == FILENODE)
+    {
+        return;
+    }
+    for (auto& childPair : l.children)
+    {
+        collectAllFingerprints(fingerprints, *childPair.second);
+    }
+}
+
+// Assigns `fa`'s fs ID to the local node from `fingerprints` that matches the fingerprint.
+// If there are multiple matches the node that's best matching the given preferred path is used.
+void assignFilesystemId(FileSystemAccess& fsaccess, FileAccess& fa, handlelocalnode_map& fsidnodes,
+                        FingerprintMap& fingerprints, string preferredNodePath,
+                        const string& localseparator, const set<string>& paths = {})
+{
+    if (!fa.fsidvalid)
+    {
+        LOG_err << "Invalid fs id";
+        return;
+    }
+
+    const auto fsId = fa.fsid;
+    FileFingerprint ffp;
+    computeFingerprint(ffp, fsaccess, fa, paths);
+    if (!ffp.isvalid)
+    {
+        return;
+    }
+
+    const auto nodeRange = fingerprints.equal_range(ffp);
+    const auto nodeCount = std::distance(nodeRange.first, nodeRange.second);
+    if (nodeCount <= 0)
+    {
+        return;
+    }
+    else if (nodeCount == 1)
+    {
+        nodeRange.first->second->setfsid(fsId);
+        fingerprints.erase(nodeRange.first);
+    }
+    else
+    {
+        // We're assigning `fa.fsid` to the node that is the best match to `preferredNodePath`.
+        // If `preferredNodePath` is in debris we're only assigning to a node that's also in debris.
+        auto bestNodeIt = nodeRange.first;
+        int bestScore = -1;
+
+        for (auto nodeIt = nodeRange.first; nodeIt != nodeRange.second; ++nodeIt)
+        {
+            string nodePath;
+            nodeIt->second->getlocalpath(&nodePath, false);
+
+            const auto score = computeReversePathMatchScore(preferredNodePath, nodePath, localseparator);
+
+            if (score > bestScore)
+            {
+                bestScore = score;
+                bestNodeIt = nodeIt;
+            }
+        }
+
+        bestNodeIt->second->setfsid(fsId);
+        fingerprints.erase(bestNodeIt);
+    }
```

Comment:
I might foresee some possible issue here. 
If I haven't misunderstood the idea here: 
in case there are more than one file with the same fingerprint as me, I will chose the one closer to my location, take it to dance and remove it from the list of available fingerprints.
But, what if that was a perfect match for another file? I might be stealing it from the rightful node. Aren't I?

Reply:
Thanks for sticking your head into this :) 
The thing is we don't know which node is the best match for fa's fingerprint. So when we have multiple nodes with the same fingerprint we pick the one that best matches fa's path. It's simply a heuristic that I thought applies in most circumstances but may sometimes be wrong (very rarely I think because fingerprints have to match). If the fingerprints on disk changed the fsid will simply stay invalid and reassigned later in the sync algo and the file considered new and re-transferred. 
I will update the comments in this functions (the debris one is stale).